### ISSUE TITLE
Evaluate deactivating or strongly reducing image processing in development mode

### ISSUE THREAD

**Issue by nkuehn** [commercetools/commercetools-docs-kit#677]
Starting a dev server takes very long and even much longer in sites with large numbers of images (half of the time or more can happen, 10 minutes too). 

Since image optimization is purely a performance optimization for production users it could be completely omitted locally.  

https://www.gatsbyjs.com/plugins/gatsby-remark-images/  and the documentation of its dependent plugins did not show an obvious way to achieve that effect.  If possible, at least being able to apply the fastest possible settings for production could help - would require some trial and error though whether it really yields an effect.

**Comment by nkuehn** [comment]
For at least a little improvement we could set `disableBgImage = true` and pass an empty array of `srcSetBreakpoints`  in dev mode.  Changing quality would not really change something because the images would be processed and rescaled in any case.   What would make a real difference would be to be able to disable PNG optimization completely. 

The documentation of the underlying low-level `gatsby-plugin-sharp` is not really consistent and clear, but I understand  that you typically don't have to directly configure it but that you can add explicit configuration which actually allows overriding things like PNG compression speed :  https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-sharp/README.md#gatsby-plugin-sharp

**Comment by nkuehn** [comment]
the docs are not as explicit but effectively sharp is plugging all options over each other (generalArgs, plugInOptions etc) - so it could be worth a try setting `pngCompressionSpeed = 10` in a gatsby config to see. 

https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-sharp/src/plugin-options.js#L21
https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-sharp/src/plugin-options.js#L79

**Comment by emmenko** [comment]
FYI: Similar feature request / issue: https://github.com/gatsbyjs/gatsby/issues/24822

**Comment by nkuehn** [comment]
ideas to try in our config:

```js
....
plugins: [
.... 
  {
    resolve: `gatsby-plugin-sharp`,
    options: process.env.NODE_ENV !== 'production' ? {} : {
              pngCompressionSpeed: 10,  // "10 is eight times faster than default" -> pngquant
              pngCompressionLevel:  4, // default 9 -> zlib compression afterwards
              jpegProgressive: false,  // minimal change only
              base64: false,  // hoping it skips the 20px blur-up thumbs
              base64Width: 0  // even if it does the thumbs, a zero size one could lead pngquant into a fast path. 
    },
  },

....

{
              resolve: 'gatsby-remark-images',
              options: {
                maxWidth: 754,
                loading: 'auto',
                srcSetBreakpoints: process.env.NODE_ENV !== 'production' ? [754, 1508] : [],
                showCaptions: ['title'],
              },
            },

....


]
```

**Comment by emmenko** [comment]
Tried that

```
width has to be a positive int larger than zero (> 0), now it's 0
```

I think it refers to the `base64Width`

**Comment by nkuehn** [comment]
A few more trial & error notes

 - removing  the 'gatsby-transformer-sharp" plugin reduces the number of processed images at startup a lot (34 to 5) but the images in the release note list don't work any more - the hook we use to provide that image data triggers processing all images independent of whether we need them or not.   
   - this is asking for a decision on whether we really need images in the release note short form and if yes, whether they can be separated into an own folder to reduce the impact. 
- emptying the array of `srcSetBreakpoints` just leads to more image processing (seems to fall back to a default list then).  not a lot to gain here since we've already reduced it to the minimal set
- `pngCompressionSpeed` is documented and implemented in some places but not actually an existing option of the underlying sharp library (or libvips which underlies sharp).  I'll file a bug with gatsyby.   https://sharp.pixelplumbing.com/api-output    (this would have been the only sort of thing that would have triggered a relevant and strong reduction).  
- `pngCompressionLevel` to control the zlib compression after image optimization has too little effect (but works)
- it is possible to reduce the number of images by reducing the srcSetBreakpoints to one outside prod builds. 

TL;DR: A little improvement is possible by further reducing the number of sizes.  Plus, we can try to optimize the images churned by the release notes list.   All beyond that requires an explicit passthrough or disable option on a lower level.

**Comment by emmenko** [comment]
Thanks for the investigation.

I was thinking more in the direction of implementing a plugin/transformer that implements the same graphql schema of the sharp/image plugin but it does not produce any optimized image.

So you can use the "custom" plugin for local development and the normal sharp/image plugins for production.

**Comment by nkuehn** [comment]
(I updated the above comment after trial runs)

@emmenko  I tend to not invest into custom hacks for this.  I found one actual incremental improvement (PR follows),  will push for a decision concerning the release notes images and would otherwise rather increase the pressure at gatsbyJS.

**Comment by nkuehn** [comment]
closing for now, we did what we could without hacking gatsby
